src/arch/arm/64/kernel/vspace.c

```diff
@@ -135,6 +135,39 @@ static word_t CONST APFromVMRights(vm_rights_t vm_rights)
     }
 }
 
+#ifndef CONFIG_ARM_HYPERVISOR_SUPPORT
+static inline CONST word_t pte_get_AP(pte_t pte)
+{
+    assert(pte_is_page_type(pte));
+    switch (pte_get_pte_type(pte)) {
+    case pte_pte_4k_page:
+        return pte_pte_4k_page_get_AP(pte);
+
+    case pte_pte_page:
+        return pte_pte_page_get_AP(pte);
+
+    default:
+        return 0;
+    }
+}
+
+static vm_rights_t CONST vmRightsFromPTE(pte_t pte)
+{
+    word_t access_perms = pte_get_AP(pte);
+    switch (access_perms) {
+    case 0:
+    case 2:
+        return VMKernelOnly;
+    case 1:
+        return VMReadWrite;
+    case 3:
+        return VMReadOnly;
+    default:
+        fail("Invalid AP bit");
```

Comment:
The compiler will complain, though, because it doesn't know that. The way it's written, the default clause is trying to communicate what you're saying -- that these are all cases, because there are only 2 bits.

Reply:
In that case it makes more sense to only have the 1 and 3 cases and make the rest default.